linkerd/tonic-watch/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+    type Future = futures::future::BoxFuture<'static, Result<OuterRsp<U>>>;
+
+    #[inline]
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<()>> {
+        // The client is cloned into the response future and driven to readiness.
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn call(&mut self, t: T) -> Self::Future {
+        Box::pin(self.clone().spawn_watch(t))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use linkerd_error::{recover, Error};
+    use linkerd_stack::{layer::Layer, MapErrLayer};
+    use tokio::{sync::mpsc, time};
+    use tokio_stream::wrappers::{IntervalStream, ReceiverStream};
+    use tower_test::mock;
+
+    #[tokio::test(flavor = "current_thread")]
+    async fn watch_reconnect() {
+        let _trace = linkerd_tracing::test::trace_init();
+
+        time::pause();
+
+        let (mock, mut handle) = mock::pair::<(), InnerRsp<u16>>();
+        let map_err = MapErrLayer::new(|e: Error| match e.downcast_ref::<tonic::Status>() {
+            Some(status) => tonic::Status::new(status.code(), status.to_string()),
+            None => tonic::Status::internal(e.to_string()),
+        });
+        let watch = StreamWatch::new(recover::Immediately::default(), map_err.layer(mock));
+
+        handle.allow(1);
+        let (tx0, rx0) = mpsc::channel::<Result<u16>>(3);
+        let send_req = handle.next_request().map(move |req| {
+            let ((), rsp) = req.unwrap();
+            rsp.send_response(tonic::Response::new(Box::pin(ReceiverStream::new(rx0))))
+        });
+        let (_, _, rx) = tokio::join!(tx0.send(Ok(123u16)), send_req, watch.spawn_watch(()));
+        let rx = rx.unwrap().into_inner();
+
+        assert_eq!(*rx.borrow(), 123);
+
+        tx0.send(Err(tonic::Status::ok("disconnect")))
+            .await
+            .unwrap();
+
+        handle.allow(1);
+        let (tx1, rx1) = mpsc::channel(3);
+        let send_req = handle.next_request().map(move |req| {
+            let ((), rsp) = req.unwrap();
+            rsp.send_response(tonic::Response::new(Box::pin(ReceiverStream::new(rx1))))
+        });
+        let (_, _) = tokio::join!(tx1.send(Ok(345u16)), send_req);
+
+        // We need to give the background task an opportunity to process the update so this yields
+        // control. The actual sleep duration is unimportant (and time is mocked, anyway).
+        time::sleep(time::Duration::from_secs(1)).await;
+
+        assert_eq!(*rx.borrow(), 345);
+    }
+
+    #[tokio::test(flavor = "current_thread")]
+    async fn block_on_initial_failure() {
+        let _trace = linkerd_tracing::test::trace_init();
+
+        time::pause();
+
+        let (mock, mut handle) = mock::pair::<(), InnerRsp<u16>>();
+        let map_err = MapErrLayer::new(|e: Error| match e.downcast_ref::<tonic::Status>() {
+            Some(status) => tonic::Status::new(status.code(), status.to_string()),
+            None => tonic::Status::internal(e.to_string()),
+        });
```

Comment:
nit/tioli: this is used in both tests, it could probably be stuck in a function? i don't really care though.